querydsl-jpa/src/main/java/com/querydsl/jpa/impl/AbstractJPAQuery.java

```diff
@@ -78,6 +78,12 @@ public AbstractJPAQuery(EntityManager em, JPQLTemplates templates, QueryMetadata
     @Override
     public long fetchCount() {
         try {
+            if (getMetadata().getGroupBy().size() > 1 || getMetadata().getHaving() != null) {
+                logger.warn("Fetch count is computing count in memory!");
+                Query query = createQuery(null, false);
```

Comment:
I think we can convert hql to native sql and add select count(1) , my solution example ：
 https://github.com/querydsl/querydsl/issues/2504#issuecomment-666888659

Reply:
That approach works only with Hibernate - not plain JPA, and only if there is a one-to-one mapping between your HQL parameters and SQL parameters, which isn't always the case (for example when a named parameter is reused in the query).

If you need reliable count queries I encourage you to use Blaze-Persistence, which has an awesome mechanism of converting any query to its count query. It also integrates with Querydsl (the `blaze-persistence-querydsl` module): https://persistence.blazebit.com/documentation/1.5/core/manual/en_US/index.html#querydsl-integration .

Using a BlazeJPAQuery instead of a JPAQuery both `fetchResults` and `fetchCount` work reliably again.